src/engine/renderer/glsl_source/computeLight_fp.glsl

```diff
@@ -235,36 +234,41 @@ void computeDynamicLights( vec3 P, vec3 normal, vec3 viewDir, vec4 diffuse, vec4
 #endif // !USE_REFLECTIVE_SPECULAR
 {
 	vec2 tile = floor( gl_FragCoord.xy * ( 1.0 / float( TILE_SIZE ) ) ) + 0.5;
+	
+	// NOT the amount of lights that can actually be put into each layer of the lighttile texture
+	uint globalLightsPerLayer = min( ( u_numLights + numLayers - 1 ) / numLayers, 16 );
 
-	#if defined(r_showLightTiles)
-		float numLights = 0.0;
-	#endif
-
-	for( int layer = 0; layer < numLayers; layer++ ) {
+	uint lightCount = 0;
+	for( uint layer = 0; layer < numLayers; layer++ ) {
 		idxs_t idxs = fetchIdxs( tileScale * vec3( tile, float( layer ) + 0.5 ), u_LightTilesInt );
-		for( int i = 0; i < lightsPerLayer; i++ ) {
-			int idx = numLayers * nextIdx( idxs ) + layer;
 
-			if( idx >= u_numLights ) {
+		uint lightOffset = layer * globalLightsPerLayer;
+		uint layerLightCount = lightOffset < u_numLights ? min( min( u_numLights - lightOffset, globalLightsPerLayer ), lightsPerLayer ) : 0;
+		for( uint i = 0; i < layerLightCount; i++ ) {
+			uint idx = nextIdx( lightCount, idxs );
+
+			if( idx == 0 ) {
 				break;
 			}
+
+			/* Light IDs are stored relative to the layer
+			Subtract 1 because 0 means there's no light */
+			idx += ( layer * globalLightsPerLayer ) - 1;
 	  
 			#if defined(USE_REFLECTIVE_SPECULAR)
 				computeDynamicLight( idx, P, normal, viewDir, diffuse, material, color, u_EnvironmentMap0, u_EnvironmentMap1 );
 			#else // !USE_REFLECTIVE_SPECULAR
 				computeDynamicLight( idx, P, normal, viewDir, diffuse, material, color );
 			#endif // !USE_REFLECTIVE_SPECULAR
 
-			#if defined(r_showLightTiles)
-				numLights++;
-			#endif
+			lightCount++;
 		}
 	}
 
 	#if defined(r_showLightTiles)
-		if ( numLights > 0.0 ) {
-			color = vec4( numLights / ( lightsPerLayer * numLayers ), numLights / ( lightsPerLayer * numLayers ),
-				numLights / ( lightsPerLayer * numLayers ), 1.0 );
+		if ( lightCount > 0.0 ) {
```

Comment:
No need to compare with a float if it's a uint I guess.

Reply:
Right, done.